enterprise/users/src/main/java/io/crate/metadata/UsersPrivilegesMetaData.java

```diff
@@ -65,6 +68,100 @@ public UsersPrivilegesMetaData(Map<String, Set<Privilege>> usersPrivileges) {
+     *
+     * NOTE: the provided privileges collection will be mutated
+     *
+     * @return the number of affected privileges (doesn't count no-ops eg. granting a privilege a user already has)
+     */
+    public long applyPrivileges(Collection<String> userNames, Iterable<Privilege> newPrivileges) {
+        long affectedPrivileges = 0L;
+        for (String userName : userNames) {
+            affectedPrivileges += applyPrivilegesToUser(userName, newPrivileges);
+        }
+        return affectedPrivileges;
+    }
+
+    private long applyPrivilegesToUser(String userName, Iterable<Privilege> newPrivileges) {
+        Set<Privilege> userPrivileges = usersPrivileges.get(userName);
+        // privileges set is expected, it must be created on user creation
+        assert userPrivileges != null : "privileges must not be null for user=" + userName;
+
+        long affectedCount = 0L;
+        for (Privilege newPrivilege : newPrivileges) {
+            Iterator<Privilege> iterator = userPrivileges.iterator();
+            boolean userHadPrivilegeOnSameObject = false;
+            while (iterator.hasNext()) {
+                Privilege userPrivilege = iterator.next();
+                if (userPrivilege.onSameObject(newPrivilege)) {
+                    userHadPrivilegeOnSameObject = true;
+                    if (newPrivilege.state().equals(State.REVOKE)) {
+                        iterator.remove();
+                        affectedCount++;
+                        break;
+                    } else {
+                        // we only want to process a new GRANT/DENY privilege if the user doesn't already have it
+                        if (userPrivilege.equals(newPrivilege) == false) {
+                            iterator.remove();
+                            userPrivileges.add(newPrivilege);
+                            affectedCount++;
+                        }
+                        break;
+                    }
+                }
+            }
+
+            if (userHadPrivilegeOnSameObject == false) {
+                // revoking a privilege that was not granted is a no-op
+                if (newPrivilege.state().equals(State.REVOKE) == false) {
+                    affectedCount++;
+                    userPrivileges.add(newPrivilege);
+                }
+            }
+        }
+
+        return affectedCount;
+    }
+
+    public long transferTablePrivileges(String sourceIdent, String targetIdent) {
+        long affectedPrivileges = 0L;
+        for (Set<Privilege> privileges : usersPrivileges.values()) {
+            Iterator<Privilege> privilegeIterator = privileges.iterator();
+            while (privilegeIterator.hasNext()) {
+                Privilege privilege = privilegeIterator.next();
+                if (privilege.clazz().equals(Privilege.Clazz.TABLE) == false) {
+                    continue;
+                }
+
+                if (privilege.ident().equals(sourceIdent)) {
```

Comment:
TABLE privileges always have an ident. But it might be a good idea to check anyway as currently it's an implicit check